### DISCUSSION TITLE
MPI run failure with petsc honoured

### DISCUSSION THREAD

**Discussion by haldaas** [firedrakeproject/firedrake#2361] [Firedrake support] [answered]
Hello,
I am trying to install firedrake on my linux machine using already installed petsc (latest version). Installation completes well. However, the installation check results in 3 failures. The simple Helmholtz problem presented in the documentation runs sequentially fine. However, a segfault occurs in parallel right in the beginning of the code (when creating the mesh).
I attach here the configure logs for petsc and firedrake.
[configure.log](https://github.com/firedrakeproject/firedrake/files/8141172/configure.log)
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/8141173/firedrake-install.log)
[make.log](https://github.com/firedrakeproject/firedrake/files/8141176/make.log)

**Comment by wence-** [comment, marked as answer]
Can you show the backtrace from the segfault? I think @ksagiyam had noticed this with a recent PETSc update too, so might be working of finding out what is going on.

**Reply by ksagiyam** [reply]
I think I've almost figured out what's going on. I think `DMPlexDistribute()` now happens by default (probably after https://gitlab.com/petsc/petsc/-/merge_requests/4843) when we call `plex.setFromOptions()` at https://github.com/firedrakeproject/firedrake/blob/4f6198461e2b96229a6aac51e4de6ab93c487d57/firedrake/mesh.py#L718, after which we explicitly call `plex.distribute()` using Chaco by default, which is a serial partitioner. The problem goes away if we run with `-petscpartitioner_type "ptscotch"`.

**Reply by haldaas** [reply]
I don't think it's very informative but here is the output.
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple MacOS to find memory corruption errors
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run
[0]PETSC ERROR: to get more information on the crash.
[0]PETSC ERROR: Run with -malloc_debug to check if memory corruption is causing the crash.
application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0

**Reply by wence-** [reply]
> I think I've almost figured out what's going on. I think `DMPlexDistribute()` now happens by default (probably after https://gitlab.com/petsc/petsc/-/merge_requests/4843) when we call `plex.setFromOptions()` at
> 
> https://github.com/firedrakeproject/firedrake/blob/4f6198461e2b96229a6aac51e4de6ab93c487d57/firedrake/mesh.py#L718
> 
> , after which we explicitly call `plex.distribute()` using Chaco by default, which is a serial partitioner. The problem goes away if we run with `-petscpartitioner_type "ptscotch"`.

Since right now we always do the distribution manually, we should wrap `DMPlexDistributeSetDefault` and `DMPlexDistributeGetDefault` and just call `dmplex.distributeSetDefault(False)` in the `Mesh` constructor before `setFromOptions`

**Reply by ksagiyam** [reply]
Thanks. Found that those functions were already wrapped. Updated https://github.com/firedrakeproject/firedrake/pull/2282.

**Reply by wence-** [reply]
@haldaas I think we have fixed the problem, could you run `firedrake-update` and try parallel tests again?

**Reply by haldaas** [reply]
All pass correctly. Thank you very much!

**Reply by wence-** [reply]
Great, thanks.
